### ISSUE TITLE
instance/node level variables

### ISSUE THREAD

**Issue by SunRunAway** [pingcap/tidb#19332]
#### tidb-server

- MySQL semantically supports `SESSION` and `GLOBAL` scoped variables, but not `INSTANCE` (for MySQL instance and global are the same thing).
- TiDB makes MySQL `GLOBAL` variables cluster-wide, but instance-level variables have a strange behavior. Some variables are actually instance-level but shown as a session-level variable, like [`tidb_slow_log_threshold`](https://docs.pingcap.com/tidb/dev/system-variables#tidb_slow_log_threshold)
- some configuration (but not all) is available in both the config file and via a `SESSION` variable. Because the config file uses a heirachy and MySQL doesn't, the names are not always obvious. Such as `slow-threshold` in the config file and `tidb_slow_log_threshold`. This confuses users.
- The instance-level variables are not persistent, which means `tidb_slow_log_threshold` takes `slow-threshold` as default value after `tidb-server` restarts each time.

#### tikv-server

* the current `set config tikv log.level="info"` will change all server's config file and reload, but not carry out when a future `tikv-server` joins.
* `set config "127.0.0.1:20180" log.level="info"` will change `127.0.0.1:20180`'s config file and reload.

### Proposal

1. Extend the system variables system to provide `INSTANCE` scope. The standard syntax will be extended to provide`SET {GLOBAL,INSTANCE,SESSION} setting='value'` AND `SHOW {GLOBAL,INSTANCE,SESSION} VARIABLES [LIKE 'setting']`. Semantics described below.
2. Provide an information_schema view of the cluster's `GLOBAL` and `INSTANCE` scoped variables. This will be used by TiDB Dashboard for managing variables and alerting suspicious instances. Semantics described further below.
3. Start to standardize data source location - ~if a configuration is controlled by a variable, it should not be controlled by the config file, vice versa.~ (lets discuss this point. I think the server should be a super-set) For example:
    - each slow log threshold of a `tidb-server` should be controlled by the instance-level variable `tidb_slow_log_threshold`.
    - `tikv_log_level` should be controlled by a variable, and should not be shown in tikv's config file. 
        + When a future `tikv-server` joins, it is applied by the global `tikv_log_level`.
        + Because there is instance scope, we can change the instance `tikv_log_level` for debugging and a canary deployment.
    - the [`pessimistic.enable`](https://docs.pingcap.com/tidb/dev/tidb-configuration-file#enable-3) should be a session/instance/global variable and not in the config file
    - the [`tmp-storage-path`](https://docs.pingcap.com/tidb/dev/tidb-configuration-file#tmp-storage-path) should stay instance scope ~in the config file because different instances may use entirely different temporary storage paths and it is hard to hot reload.~ (lets discuss this point. I think it should be instance scoped and still accessible in the server).

#### Instance Scope Semantics

Variables that are currently scoped both `GLOBAL` and `SESSION` have the semantic that `SESSION` inherits from `GLOBAL` when the `SESSION` starts. Any further change to `GLOBAL` does not propagate to `SESSION`, and `SESSION` is lost when the session ends.

The semantics for `INSTANCE` are proposed as:

1. If a variable is scoped as `INSTANCE`, it will also appear in the output of `SHOW GLOBAL VARIABLES`. Changing an `INSTANCE` variable will also "appear" to change the `GLOBAL` variable (as it appears in `SHOW GLOBAL VARIABLES`), but will not persist to the rest of the cluster. The fact that it is not a "true global" will be able to be seen in `information_schema`. This helps prevent confusion from DBAs that are acclimatized by the inheritance rules, and will be confused be a behavior change.

2. Changes to `INSTANCE` variables will persist restarts.

3. If an `INSTANCE` variable has been set, it will take precedence over any changes elsewhere in the cluster to `SET GLOBAL` and overwrite that parameter. For the global parameter to apply, the command `RESET INSTANCE <setting>` will need to be called.
  * An exception to this, is if `SET GLOBAL` was called on this instance. In which case, to respect MySQL semantics it should imply `RESET INSTANCE <setting>` on that specific instance.

#### Information_schema view

This can be similar to `performance_schema.variables_info` from MySQL 8.0. It does not need to show sessions, only settings sourced from configuration files, instance, global.

### Migration Plan

We need to carefully design the migration procedure if this proposal is approved. One option is that we provide a bootstrap procedure in tidb-server. When the bootstrap happens, it collects the existing configuration in the config file, set them as a global variable if all of them are in common, and set instance variable if some instance is special. At last, the configuration in the config file is deleted by the bootstrap.

For example, migrate `log.level` from tikv config file to an instance/global variable,

1. The bootstrap collects all the configurations from all living tikv-server's config file.
2. Set the variable,
   1. if the values of `log.level` from config files are all the same, set it as `tikv.log.level` 's global variable.
   2. set instance variable if some instance's value is special
3. Remove all `log.level` from the config file, and leave a comment in place to tell users what happened.
4. If some configuration is still shown in the config file, `tidb-server` should warn users when it starts and tells users that they do not take effect.

This migration plan is also suitable for TiUP and TiDB-Operator. Consider the initial deployment steps for TiUP and TiDB-Operator, as well as upgrading steps for TiUP and TiDB-Operator.
For TiUP / TiDB Operator initial deployment, we'd better provide a new config set, for example, `initialGlobalVariables`, to help users automatically set these global variables after initial deployment.

For TiUP / TiDB Operator upgrading, these no-longer-exist config items should be removed from the config. For TiUP, this kind of removal can be automatically done by a newer TiUP cluster binary. For tidb-operator, maybe we should warn users about this.

### Name remapping (From @nullnotnil )

What I would propose is that config file variables have a mapping to mysql-like names, so they can naturally appear in this list https://docs.pingcap.com/tidb/stable/system-variables and use MySQL sysvar setting syntax. In most cases, this can be done by converting a `.` to `_`. Thus the current syntax of `set config` (TiDB proprietary) is not required. It can just be MySQL syntax :

```sql
set config tikv log.level="info";
# can become
set global tikv_log_level="info";
```

There might be some variables that need editing, for example:

```sql
set config pessimistic.enable = 1;
# should becomes
SET GLOBAL tidb_enable_pessimistic_transactions = 1;
```

(This helps respect an existing convention of `tidb_enable%` sysvars.)

### Documentation changes

1. An introduction page for session/instance/global variables is necessary.
2. We should rearrange the config file and variables.
3. Because most configurations can be migrated to variables, [`Modify Configuration Online`](https://docs.pingcap.com/tidb/dev/dynamic-config) is discussable to retire.

### reference: 

1. The current implementation https://docs.pingcap.com/tidb/dev/dynamic-config
2. SET CLUSTER SETTING(CockroachDB) https://www.cockroachlabs.com/docs/v20.1/set-cluster-setting
3. https://dev.mysql.com/doc/refman/8.0/en/set-variable.html

**Comment by breeswish** [comment]
@lonng @nullnotnil @zz-jason @qw4990 PTAL

**Comment by breeswish** [comment]
I think we also need to consider the initial deployment steps for TiUP and TiDB-Operator, as well as upgrading steps for TiUP and TiDB-Operator.

For TiUP / TiDB Operator initial deployment, we'd better provide a new config cluster, for example, `initialGlobalVariables`, to help user automatically set these global variables after initial deployment.

For TiUP / TiDB Operator upgrading, these no-longer-exist config items should be removed from the config. For TiUP, this kind of removal can be automatically done by a newer TiUP cluster binary. For tidb-operator, maybe we should warn users about this.

**Comment by SunRunAway** [comment]
> I think we also need to consider the initial deployment steps for TiUP and TiDB-Operator, as well as upgrading steps for TiUP and TiDB-Operator.
> 
> For TiUP / TiDB Operator initial deployment, we'd better provide a new config set, for example, `initialGlobalVariables`, to help user automatically set these global variables after initial deployment.
> 
> For TiUP / TiDB Operator upgrading, these no-longer-exist config items should be removed from the config. For TiUP, this kind of removal can be automatically done by a newer TiUP cluster binary. For tidb-operator, maybe we should warn users about this.

Thanks, I've added your idea into the Migration Plan section.

**Comment by nullnotnil** [comment]
There is an existing proposals or instance scoped variables here: https://github.com/pingcap/tidb/issues/9298

Can these two be merged?

**Comment by SunRunAway** [comment]
> There is an existing proposals or instance scoped variables here: #9298
> 
> Can these two be merged?

Sorry I hadn't seen the existing proposal before. We can discuss here because it has a detailed plan for migrating items in config file to instance/global variables.

**Comment by nullnotnil** [comment]
SGTM. I like the migration plan, and the details all make sense. What I would propose is that config file variables have a mapping to mysql-like names, so they can naturally appear in this list https://docs.pingcap.com/tidb/stable/system-variables and use MySQL sysvar setting syntax. In most cases, this can be done by converting a `.` to `_`. Thus:

```sql
set config tikv log.level="info";
# can become
set global tikv_log_level="info";
```

There might be some variables that need editing, for example:

```sql
set config pessimistic.enable = 1;
# should becomes
SET GLOBAL tidb_enable_pessimistic_transactions = 1;
```

(This helps respect an existing convention of `tidb_enable%` sysvars.)

I am happy to review and create the mapping if we agree in principle.

**Comment by zz-jason** [comment]
I think it's duplicated with this issue: https://github.com/pingcap/tidb/issues/9298

**Comment by SunRunAway** [comment]
> SGTM. I like the migration plan, and the details all make sense. What I would propose is that config file variables have a mapping to mysql-like names, so they can naturally appear in this list https://docs.pingcap.com/tidb/stable/system-variables and use MySQL sysvar setting syntax. In most cases, this can be done by converting a `.` to `_`. Thus:
> 
> ```sql
> set config tikv log.level="info";
> # can become
> set global tikv_log_level="info";
> ```
> 
> There might be some variables that need editing, for example:
> 
> ```sql
> set config pessimistic.enable = 1;
> # should becomes
> SET GLOBAL tidb_enable_pessimistic_transactions = 1;
> ```
> 
> (This helps respect an existing convention of `tidb_enable%` sysvars.)
> 
> I am happy to review and create the mapping if we agree in principle.

LGTM. I've added the `Name remapping` section in the issue description.

**Comment by breeswish** [comment]
> > SGTM. I like the migration plan, and the details all make sense. What I would propose is that config file variables have a mapping to mysql-like names, so they can naturally appear in this list https://docs.pingcap.com/tidb/stable/system-variables and use MySQL sysvar setting syntax. In most cases, this can be done by converting a `.` to `_`. Thus:
> > ```sql
> > set config tikv log.level="info";
> > # can become
> > set global tikv_log_level="info";
> > ```
> > 
> > 
> > There might be some variables that need editing, for example:
> > ```sql
> > set config pessimistic.enable = 1;
> > # should becomes
> > SET GLOBAL tidb_enable_pessimistic_transactions = 1;
> > ```
> > 
> > 
> > (This helps respect an existing convention of `tidb_enable%` sysvars.)
> > I am happy to review and create the mapping if we agree in principle.
> 
> LGTM. I've added the `Name remapping` section in the issue description.

What about existing config items that have "_" in the name? To support it, does it mean that we need to *hard code* all possible TiKV configuration names in TiDB?

**Comment by nullnotnil** [comment]
> What about existing config items that have "_" in the name? To support it, does it mean that we need to _hard code_ all possible TiKV configuration names in TiDB?

No I think that is a bit excessive. If you execute `show global variables` as well, they will show up (including aliases) which gets messy fast. I am working on a more detailed proposal to find cases where there is:

1. a pre-existing sysvar (no change required)
2. a MySQL sysvar that is identical in functionality (can we just offer the feature as compatibility?)
3. a config file option but not a variable (requires a mapping, using the methodology described)

**Comment by morgo** [comment]
I am going to close this as a duplicate of https://github.com/pingcap/tidb/issues/30557

The feature has been mostly implemented in TiDB 6.1 by @CbcWestwolf
